vm/natives.c

```diff
@@ -1138,5 +1138,393 @@ const AsebaNativeFunctionDescription AsebaNativeDescription_rand =
+// in another array. For example, call.pop_front(result[3:5], dq) will pop three elements
+// from a deque stored in array dq into positions 3:5 of the array result. It is the
+// programmer's responsibility to use consistent tuple sizes.
+
+// Inserting and erasing must shift existing elements in the deque. The deque_shift private
+// function replaces the followinf specific loops:
+//
+// Insert in left half: shift prefix elements left
+//		for (k = 0; k <= index_val - 1 + src_length - 1; k++)
+//			vm->variables[dest + 2 + ((dq_start + k) % dq_capacity)] = vm->variables[dest + 2 + ((dq_start + k + src_length) % dq_capacity)];
+// Insert in right half: shift prefix elements left
+//		for (k = dq_size + src_length - 1; k > index_val + src_length - 1; k--)
+//			vm->variables[dest + 2 + ((dq_start + k) % dq_capacity)] = vm->variables[dest + 2 + ((dq_start + k - src_length) % dq_capacity)];
+// Erase in left half: shift prefix elements right
+//		for (k = index_val; k > 0; k--)
+//			vm->variables[dest + 2 + ((dq_start + k) % dq_capacity)] = vm->variables[dest + 2 + ((dq_start + k - 1) % dq_capacity)];
+// Erase in right half: shift prefix elements left
+//		for (k = index_val; k < dq_size - 1; k++)
+//			vm->variables[dest + 2 + ((dq_start + k) % dq_capacity)] = vm->variables[dest + 2 + ((dq_start + k + 1) % dq_capacity)];
+
+static void deque_shift(AsebaVMState *vm, uint16 dq, uint16 dq_capacity, uint16 target, sint16 last, sint16 delta)
+{
+	for ( ; (delta < 0 ? target >= last : target <= last) ; (delta < 0 ? target-- : target++) )
+		vm->variables[dq + 2 + (target % dq_capacity)] = vm->variables[dq + 2 + ((target + delta) % dq_capacity)];
+}
+
+void AsebaNative_deqsize(AsebaVMState *vm)
+{
+	// variable pos
+	uint16 size = AsebaNativePopArg(vm);
+	uint16 dest = AsebaNativePopArg(vm);
+	
+	// variable size
+	uint16 dest_length = AsebaNativePopArg(vm);
+
+	vm->variables[size++] = vm->variables[dest++];
+}
+
+const AsebaNativeFunctionDescription AsebaNativeDescription_deqsize =
+{
+	"deque.size",
+	"report size of deque dest",
+	{
+		{  1, "size" },
+		{ -1, "dest" },
+		{ 0, 0 }
+	}
+};
+
+static void _AsebaNative_deqget(AsebaVMState *vm, uint16 dest, uint16 src, uint16 index_val, uint16 dest_length, uint16 src_length)
+{
+	// infer deque parameters
+	uint16 dq_start = vm->variables[src + 1];
+	uint16 dq_capacity = src_length - 2;
+
+	// copy elements from deque
+	uint16 i;
+
+	for (i = 0; i < dest_length; i++)
+	{
+		vm->variables[dest++] = vm->variables[src + 2 + ((dq_start + index_val + i) % dq_capacity)];
+	}
+}
+
+void AsebaNative_deqget(AsebaVMState *vm)
+{
+	// variable pos
+	uint16 dest = AsebaNativePopArg(vm);
+	uint16 src = AsebaNativePopArg(vm);
+	uint16 index = AsebaNativePopArg(vm);
+	uint16 index_val = vm->variables[index];
+	
+	// variable size
+	uint16 dest_length = AsebaNativePopArg(vm);
+	uint16 src_length = AsebaNativePopArg(vm);
+	
+	// Check for exception
+	if (vm->variables[index] < 0 || index_val > src_length - 2 - 1
```

Comment:
Because I want to throw an exception if the supplied argument is < 0, while declaring `index_val` unsigned as it should be. Since an exception is raised in `index_val` is larger than the array, the silent limitation of `index_val` to [0,32767) isn't a constraint.